staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/controller.go

```diff
@@ -432,6 +411,124 @@ func (c *celAdmissionController) Validate(
 	return nil
 }
 
+// Returns objects to use to evaluate the policy
+func (c *celAdmissionController) collectParams(
+	paramKind *v1alpha1.ParamKind,
+	info paramInfo,
+	paramRef *v1alpha1.ParamRef,
+	namespace string,
+) ([]runtime.Object, error) {
+	// If definition has paramKind, paramRef is required in binding.
+	// If definition has no paramKind, paramRef set in binding will be ignored.
+	var params []runtime.Object
+
+	// Make sure the param kind is ready to use
+	if paramKind != nil && (paramRef != nil) {
+		if info.controller == nil {
+			return nil, fmt.Errorf("paramKind kind `%v` not known",
+				paramKind.String())
+		}
+
+		// If the param informer for this admission policy has not yet
+		// had time to perform an initial listing, don't attempt to use
+		// it.
+		timeoutCtx, cancel := context.WithTimeout(c.policyController.context, 1*time.Second)
+		defer cancel()
+
+		if !cache.WaitForCacheSync(timeoutCtx.Done(), info.controller.HasSynced) {
+			return nil, fmt.Errorf("paramKind kind `%v` not yet synced to use for admission",
+				paramKind.String())
+		}
+	}
+
+	// Find params to use with policy
+	switch {
+	case paramKind == nil:
+		// ParamKind is unset. Ignore any globalParamRef or namespaceParamRef
+		// setting.
+		return []runtime.Object{nil}, nil
+	case paramRef == nil:
+		// Policy ParamKind is set, but binding does not use it.
+		// Validate with nil params
+		return []runtime.Object{nil}, nil
+	case paramRef != nil && len(paramRef.Name) > 0 && paramRef.Selector == nil:
+		var param runtime.Object
+		var err error
+
+		if info.scope.Name() == meta.RESTScopeRoot.Name() {
+			if len(paramRef.Namespace) > 0 {
+				return nil, fmt.Errorf("paramRef.namespace must not be provided for a cluster-scoped `paramKind`")
+			}
+
+			param, err = info.controller.Informer().Get(paramRef.Name)
+		} else {
+			paramsNamesace := namespace
+			if len(paramRef.Namespace) > 0 {
+				paramsNamesace = paramRef.Namespace
+			} else if len(paramsNamesace) == 0 {
+				paramsNamesace = v1.NamespaceDefault
+			}
+			param, err = info.controller.Informer().Namespaced(paramsNamesace).Get(paramRef.Name)
+		}
+
+		switch {
+		case err == nil:
+			params = []runtime.Object{param}
+		case k8serrors.IsNotFound(err):
+			// Param not yet available. User may need to wait a bit
+			// before being able to use it for validation.
+			//
+			// Do nothing, apply ParameterNotFoundAction below
+		case k8serrors.IsInvalid(err):
+			// Param mis-configured
+			// require to set namespace for namespaced resource
+			// and unset namespace for cluster scoped resource
+			return nil, err
+		}
+	case paramRef != nil && len(paramRef.Name) == 0 && paramRef.Selector != nil:
+		// Check that param kind is namespaced
+		if info.scope != nil && info.scope.Name() != meta.RESTScopeNamespace.Name() {
+			return nil, errors.New("NamespaceParamRef may only be used on a binding whose ParamKind is namespace-scoped")
```

Comment:
nit: error message includes removed field `NamespaceParamRef` would cause confusion.

Reply:
changed error to:

```
paramRef.namespace must not be provided for a cluster-scoped `paramKind`
```